diff --git a/pkgs/tools/archivers/7zz/default.nix b/pkgs/tools/archivers/7zz/default.nix
index ca8f8366f7f6b..d23ec4af02ee4 100644
--- a/pkgs/tools/archivers/7zz/default.nix
+++ b/pkgs/tools/archivers/7zz/default.nix
@@ -77,6 +77,7 @@ stdenv.mkDerivation (finalAttrs: {
       "CXX=${stdenv.cc.targetPrefix}c++"
     ]
     ++ lib.optionals useUasm [ "MY_ASM=uasm" ]
+    ++ lib.optionals (!useUasm && stdenv.hostPlatform.isx86) [ "USE_ASM=" ]
     # We need at minimum 10.13 here because of utimensat, however since
     # we need a bump anyway, let's set the same minimum version as the one in
     # aarch64-darwin so we don't need additional changes for it
