diff --git a/pkgs/applications/misc/bambu-studio/default.nix b/pkgs/applications/misc/bambu-studio/default.nix
index 481f365717064..82b6de9391b79 100644
--- a/pkgs/applications/misc/bambu-studio/default.nix
+++ b/pkgs/applications/misc/bambu-studio/default.nix
@@ -1,9 +1,6 @@
 {
   stdenv,
   lib,
-  openexr,
-  jemalloc,
-  c-blosc,
   binutils,
   fetchFromGitHub,
   cmake,
@@ -55,17 +52,6 @@ let
           "--enable-debug=no"
         ];
       });
-
-  openvdb' = openvdb.overrideAttrs (old: {
-    buildInputs = [
-      openexr
-      boost179
-      tbb_2021_11
-      jemalloc
-      c-blosc
-      ilmbase
-    ];
-  });
 in
 stdenv.mkDerivation rec {
   pname = "bambu-studio";
@@ -111,7 +97,7 @@ stdenv.mkDerivation rec {
     mpfr
     nlopt
     opencascade-occt_7_6
-    openvdb'
+    openvdb
     pcre
     tbb_2021_11
     webkitgtk_4_0
