diff --git a/pkgs/by-name/yo/your_spotify/package.nix b/pkgs/by-name/yo/your_spotify/package.nix
index 9ab6e61af77b7..64bbec67a65e9 100644
--- a/pkgs/by-name/yo/your_spotify/package.nix
+++ b/pkgs/by-name/yo/your_spotify/package.nix
@@ -51,6 +51,7 @@ stdenv.mkDerivation (finalAttrs: {
 
     mkdir -p $out/share/your_spotify
     cp -r node_modules $out/share/your_spotify/node_modules
+    cp -r apps $out/share/your_spotify/apps
     cp -r ./apps/server/{lib,package.json} $out
     mkdir -p $out/bin
     makeWrapper ${lib.escapeShellArg (lib.getExe nodejs)} "$out/bin/your_spotify_migrate" \
