diff --git a/nixos/tests/kanidm-provisioning.nix b/nixos/tests/kanidm-provisioning.nix
index 16e39dba729c0a..8f0ca0ec08592d 100644
--- a/nixos/tests/kanidm-provisioning.nix
+++ b/nixos/tests/kanidm-provisioning.nix
@@ -23,7 +23,7 @@ import ./make-test-python.nix (
       { pkgs, lib, ... }:
       {
         services.kanidm = {
-          package = pkgs.kanidmWithSecretProvisioning;
+          package = pkgs.kanidmWithSecretProvisioning_1_6;
           enableServer = true;
           serverSettings = {
             origin = "https://${serverDomain}";
diff --git a/nixos/tests/kanidm.nix b/nixos/tests/kanidm.nix
index b3b4e0213cbfa3..69dac0de686523 100644
--- a/nixos/tests/kanidm.nix
+++ b/nixos/tests/kanidm.nix
@@ -26,6 +26,7 @@ import ./make-test-python.nix (
       { pkgs, ... }:
       {
         services.kanidm = {
+          package = pkgs.kanidm_1_6;
           enableServer = true;
           serverSettings = {
             origin = "https://${serverDomain}";
@@ -55,6 +56,7 @@ import ./make-test-python.nix (
       { nodes, ... }:
       {
         services.kanidm = {
+          package = pkgs.kanidm_1_6;
           enableClient = true;
           clientSettings = {
             uri = "https://${serverDomain}";
diff --git a/pkgs/by-name/ka/kanidm/generic.nix b/pkgs/by-name/ka/kanidm/generic.nix
index 34e4806500862a..08ac4c490901f1 100644
--- a/pkgs/by-name/ka/kanidm/generic.nix
+++ b/pkgs/by-name/ka/kanidm/generic.nix
@@ -49,7 +49,7 @@ rustPlatform.buildRustPackage rec {
     inherit hash;
   };
 
-  KANIDM_BUILD_PROFILE = "release_nixpgs_${arch}";
+  KANIDM_BUILD_PROFILE = "release_nixpkgs_${arch}";
 
   patches = lib.optionals enableSecretProvisioning [
     "${patchDir}/oauth2-basic-secret-modify.patch"
@@ -59,7 +59,7 @@ rustPlatform.buildRustPackage rec {
   postPatch =
     let
       format = (formats.toml { }).generate "${KANIDM_BUILD_PROFILE}.toml";
-      socket_path = if stdenv.hostPlatform.isLinux then "/run/kanidm/sock" else "/var/run/kanidm.socket";
+      socket_path = if stdenv.hostPlatform.isLinux then "/run/kanidmd/sock" else "/var/run/kanidm.socket";
       profile =
         {
           cpu_flags = if stdenv.hostPlatform.isx86_64 then "x86_64_legacy" else "none";
